src/Microsoft.Performance.SDK.Runtime.NetCoreApp/Plugins/PluginsLoader.cs

```diff
@@ -0,0 +1,255 @@
+using Microsoft.Performance.SDK.Runtime.Extensibility.DataExtensions;
+using Microsoft.Performance.SDK.Runtime.Extensibility.DataExtensions.Repository;
+using Microsoft.Performance.SDK.Runtime.NetCoreApp.Discovery;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+
+namespace Microsoft.Performance.SDK.Runtime.NetCoreApp.Plugins
+{
+    /// <summary>
+    ///     Handles loading SDK plugins into isolation assembly contexts and notifying
+    ///     <see cref="IPluginsConsumer"/> listeners of all custom data sources loaded
+    ///     by plugins.
+    ///     <para/>
+    ///     Plugins are loaded by supplying either <see cref="TryLoadPlugin(string)"/> or
+    ///     <see cref="TryLoadPlugins(IEnumerable{string}, out IList{string})"/> directories
+    ///     that contain all binaries necessary for the plugins to run. The filesystem structure
+    ///     of plugins is expected to be 
+    ///         InstalledPlugins\<manager_version>\<plugin_name>\<plugin_version>
+    ///     For example, a plugin named "foo" with two versions, 1.0.0 and 2.0.0,
+    ///     may have the following folder structure:
+    ///         C:\SDK\InstalledPlugins\1.2.3\foo\1.0.0\<files>
+    ///         C:\SDK\InstalledPlugins\1.2.3\foo\2.0.0\<files>
+    ///     In this example, a specific version of "foo" can be loaded by passing in the full path of
+    ///     the "1.0.0" or "2.0.0" folder to <see cref="TryLoadPlugin(string)"/> or <see cref="TryLoadPlugins(IEnumerable{string}, out IList{string})"/>.
+    ///     Note that passing in the full path to "foo" or any of its parents will result in both
+    ///     versions of the plugin "foo" being loaded.
+    ///     <para/>
+    ///     Failure to adhere to this filesystem structure will result in plugins being loaded with
+    ///     an unknown (<c>null</c>) name and version.
+    ///     <para/>
+    ///     <see cref="IPluginsConsumer"/> listeners can register to be notified of plugins loaded in
+    ///     the future through <see cref="Subscribe(IPluginsConsumer)"/>. When a new listener subscribes,
+    ///     before <see cref="Subscribe(IPluginsConsumer)"/> returns the consumer is first notified
+    ///     of all plugins loaded prior to the subscription.
+    ///     <para/>
+    ///     All calls to loading or obtaining loaded plugins are thread-safe occur in a linearizable order. 
+    ///     Listeners are guaranteed to not miss any loaded custom data sources that get loaded concurrently
+    ///     with their <see cref="Subscribe(IPluginsConsumer)"/> call.
+    /// </summary>
+    public class PluginsLoader
+    {
+        private readonly object mutex = new object();
+
+        private readonly AssemblyExtensionDiscovery extensionDiscovery;
+
+        private readonly IPlugInCatalog catalog;
+
+        private readonly IDataExtensionRepositoryBuilder extensionRepository;
+
+        private readonly ICollection<IPluginsConsumer> subscribers;
```

Comment:
>ICollection [](start = 25, length = 11)

because this is private and it's enumerated, would it be better to just use ConcurrentSet here for performance?

Reply:
Made this a HashSet since I'm moving to using the mutex for unsubscribe